<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Загрузка-и-изучение-данных" data-toc-modified-id="Загрузка-и-изучение-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Загрузка и изучение данных</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span><ul class="toc-item"><li><span><a href="#Теоретическая-часть" data-toc-modified-id="Теоретическая-часть-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Теоретическая часть</a></span></li><li><span><a href="#Признаки-умножают-на-обратимую-матрицу.-Изменится-ли-качество-линейной-регрессии?" data-toc-modified-id="Признаки-умножают-на-обратимую-матрицу.-Изменится-ли-качество-линейной-регрессии?-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии?</a></span></li></ul></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Исследуем-качество-модели-без-преобразования" data-toc-modified-id="Исследуем-качество-модели-без-преобразования-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Исследуем качество модели без преобразования</a></span><ul class="toc-item"><li><span><a href="#С-исходными-признаками" data-toc-modified-id="С-исходными-признаками-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>С исходными признаками</a></span></li><li><span><a href="#С-отмасштабированными-признаками" data-toc-modified-id="С-отмасштабированными-признаками-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>С отмасштабированными признаками</a></span></li></ul></li><li><span><a href="#Исследуем-качество-модели-с-преобразованием" data-toc-modified-id="Исследуем-качество-модели-с-преобразованием-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Исследуем качество модели с преобразованием</a></span><ul class="toc-item"><li><span><a href="#Преобразование-матрицы" data-toc-modified-id="Преобразование-матрицы-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Преобразование матрицы</a></span></li><li><span><a href="#Качество-модели-с-исходными-признаками" data-toc-modified-id="Качество-модели-с-исходными-признаками-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Качество модели с исходными признаками</a></span></li><li><span><a href="#Качество-модели-с-отмасштабированными-признаками" data-toc-modified-id="Качество-модели-с-отмасштабированными-признаками-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Качество модели с отмасштабированными признаками</a></span></li></ul></li></ul></li><li><span><a href="#Итоговый-вывод" data-toc-modified-id="Итоговый-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Итоговый вывод</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

### Импорт библиотек

In [1]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

### Загрузка и изучение данных

In [2]:
# прочитаем файл и сохраним его в переменной df
df = pd.read_csv('/datasets/insurance.csv')

# выведем первые 5 строк на экран
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [3]:
# с полощью метода info() изучим общую информацию о данных: количество строк и столбцов, типы данных и наличие пропуков
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [4]:
df.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

Проверим, есть ли в данных дубликаты

In [5]:
df[df.duplicated() == True].count()

Пол                  153
Возраст              153
Зарплата             153
Члены семьи          153
Страховые выплаты    153
dtype: int64

Количество дубликатов - 153 строчки

Изменим тип данных в столбцах Возраст и Зарплата на целочисленный ввиду отсуствия дробных частей. Так же уменьшим обращение к памяти для остальных столбцов

In [6]:
print('Максимальный остаток от деления на 1:', (df['Зарплата'] % 1).max())
print('Средний остаток от деления на 1:', (df['Зарплата'] % 1).mean())
print('Минимальный остаток от деления на 1:', (df['Зарплата'] % 1).min())

Максимальный остаток от деления на 1: 0.999999999992724
Средний остаток от деления на 1: 0.0006000000000240107
Минимальный остаток от деления на 1: 0.0


In [7]:
df = df.astype(int, errors='ignore')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Пол                5000 non-null   int64
 1   Возраст            5000 non-null   int64
 2   Зарплата           5000 non-null   int64
 3   Члены семьи        5000 non-null   int64
 4   Страховые выплаты  5000 non-null   int64
dtypes: int64(5)
memory usage: 195.4 KB


In [8]:
df['Пол'] = pd.to_numeric(df['Пол'], downcast='integer')
df['Возраст'] = pd.to_numeric(df['Возраст'], downcast='integer')
df['Зарплата'] = pd.to_numeric(df['Зарплата'], downcast='integer')
df['Члены семьи'] = pd.to_numeric(df['Члены семьи'], downcast='integer')
df['Страховые выплаты'] = pd.to_numeric(df['Страховые выплаты'], downcast='integer')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Пол                5000 non-null   int8 
 1   Возраст            5000 non-null   int8 
 2   Зарплата           5000 non-null   int32
 3   Члены семьи        5000 non-null   int8 
 4   Страховые выплаты  5000 non-null   int8 
dtypes: int32(1), int8(4)
memory usage: 39.2 KB


Проведем корреляционный анализ и посмотрим, как переменные взаимосвязаны между собой

In [9]:
df.corr()

Пол   Возраст  Зарплата  Члены семьи  \
Пол                1.000000  0.002074  0.014910    -0.008991   
Возраст            0.002074  1.000000 -0.019093    -0.006692   
Зарплата           0.014910 -0.019093  1.000000    -0.030296   
Члены семьи       -0.008991 -0.006692 -0.030296     1.000000   
Страховые выплаты  0.010140  0.651030 -0.014963    -0.036290   

                   Страховые выплаты  
Пол                         0.010140  
Возраст                     0.651030  
Зарплата                   -0.014963  
Члены семьи                -0.036290  
Страховые выплаты           1.000000

**Вывод:**
- Количество мужчин и женщин в датасете поравну
- Возраст от 18 до 65 лет и в среднем людям 31
- Зарплата от 5300 до 79000
- Членов семьи в среднем 1, максимально 6
- Страховых выплат обычно нет, но максимально 5
- Страховые выплаты сильно коррелируют с Возрастом и имеют обратную корреляцию и Количеством членов семьи и Зарплатой

## Умножение матриц

Убедимся в том, что линейня регрессия из scikit-learn вычисляет значения по формуле:

In [10]:
features = df.drop('Страховые выплаты', axis=1)
target = df['Страховые выплаты']

In [11]:
X = np.concatenate((np.ones((features.shape[0], 1)), features), axis=1)
y = target
w = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)
display(w[1:])

array([ 7.92580563e-03,  3.57083050e-02, -1.70081903e-07, -1.35676627e-02])

In [12]:
model = LinearRegression()
model.fit(features, target)
model.coef_

array([ 7.92580563e-03,  3.57083050e-02, -1.70081903e-07, -1.35676627e-02])

**Вывод:** Коэфициенты регрессии совпадают на 100%

### Теоретическая часть

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

### Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии?

**Ответ:** не изменится.

**Обоснование:**

Заменим матрицу ***X*** на матрицу ***Z***:

$$
Z = XP
$$

где ***P*** - обратимая матрица с некими значениями, на которую может быть умножена матрица 

Заменим ***X*** на ***Z*** и вычислим, чему будет равено предсказание и вектор весов.

$$
a_{1} = Zw_{1}
$$

$$
w_{1} = (Z^T Z)^{-1} Z^T y
$$

Подставим второе уравнение в первое:

$$
a_{1} = Z(Z^T Z)^{-1} Z^T y
$$

Заменим все ***Z***:

$$
a_{1} = XP((XP)^T (XP))^{-1} (XP)^T y
$$

Воспользуемся свойством транспорнированной матрицы:

$$
(AB)^T = B^T A^T
$$

$$
a_{1} = XP(P^T X^T X P))^{-1} (XP)^T y
$$

Свойсво обратной матрицы:

$$
(AB)^{-1} = (B)^{-1} (A)^{-1}
$$

$$
a_{1} = XP(P)^{-1} (X^T X)^{-1} (P^T)^{-1} P^T X^T  y
$$

$$
P(P)^{-1} = E
$$

$$
(P^T)^{-1} (P)^T = E
$$

Умножение на единичную матрицу ничего не меняет.

$$
a_{1} = X E (X^T X)^{-1} E X^T  y = X (X^T X)^{-1} (X)^T y = Xw = a
$$


**Вывод:** значение предсказания не меняется, если умножать матрицу признаков на обратимую матрицу

## Алгоритм преобразования

**Алгоритм**

Для защиты информации на этапе поиска коэффициентов регрессии и предсказаний регрессии будем умножать матрицу признаков на обратимую матрицу , которая будет генерироваться случайным образом.

Этапы алгоритма:

1. Составление матрицы ***Y***
2. Проверка матрицы на обратимость. Вычисление детерминанта матрицы ***Y***
3. Получение матрицы преобразованных признаков ***Z=XY***
4. Применение алгоритма на преобразованных признаках ***Z***

**Обоснование**

Матрица ***Y*** должна иметь размерность ***nxn***, где n - количество признаков для регрессии. Матрица ***Z*** будет иметь туже размерность, что и матрица ***X***. 

$$
X = 
\begin{pmatrix}
1 & 2 \\
2 & 3 \\
4 & 5 \\
\end{pmatrix}
$$

$$
Y = 
\begin{pmatrix}
1 & 0 \\
2 & 3 \\
\end{pmatrix}
$$

Найдем значние ***Z***:

$$
Z = XY = 
\begin{pmatrix}
1*1+2*2 & 1*0+2*3 \\
2*1+3*2 & 2*0+3*3 \\
4*1+5*2 & 4*0+5*3 \\
\end{pmatrix}
= 
\begin{pmatrix}
5 & 6 \\
8 & 9 \\
14 & 15 \\
\end{pmatrix}
$$

После этого добавляем нулевой столбец и вводим данные в линейную регрессию

## Проверка алгоритма

Проведем исследование модели по двум направлениям:

- Исследуем качество модели без преобразования.
   - С исходными признаками
   - С отмасштабированными признаками
- Исследуем качество модели с преобразованием. 
   - С исходными признаками 
   - С отмасштабированными признаками

Разделим данные на обучающие и тестовые

In [13]:
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.25, random_state=12345)

In [14]:
print('Размер тренировочной выборки', features_train.shape)
print('Размер тестовой выборки', features_test.shape)

Размер тренировочной выборки (3750, 4)
Размер тестовой выборки (1250, 4)


### Исследуем качество модели без преобразования

#### С исходными признаками

In [15]:
model = LinearRegression()
model.fit(features_train, target_train)

R2_origin_data = r2_score(target_test, model.predict(features_test))

print("Коэффициенты - вектор w:",model.coef_)
print("R2 =", R2_origin_data)

Коэффициенты - вектор w: [ 1.79258370e-02  3.57228278e-02 -5.46002729e-07 -1.26186599e-02]
R2 = 0.4352275684083322


#### С отмасштабированными признаками

In [16]:
regressor = LinearRegression()
scaller = StandardScaler()
scaller.fit(features_train, target_train)

pipeline = Pipeline([("standard_scaller", scaller),("linear_regression", regressor)])
pipeline.fit(features_train, target_train)

R2_LR_origin_data_scaled = r2_score(target_test, pipeline.predict(features_test))

print("R2 =", R2_LR_origin_data_scaled)

R2 = 0.43522756840833376


**Вывод:** Модель ведет себя одинаково как на исходных данных, так и на отмасштабированных

### Исследуем качество модели с преобразованием

#### Преобразование матрицы

Создадим матрицу признаков:

In [17]:
features_matrix = features.values
features_matrix[:5, :]

array([[    1,    41, 49600,     1],
       [    0,    46, 38000,     1],
       [    0,    29, 21000,     0],
       [    0,    21, 41700,     2],
       [    1,    28, 26100,     0]], dtype=int32)

Создадим случайную матрицу nxn

In [18]:
n = features.shape[1]
np.random.seed(42)
random_matrix = np.random.randint(1, 10, (n,n))
random_matrix

array([[7, 4, 8, 5],
       [7, 3, 7, 8],
       [5, 4, 8, 8],
       [3, 6, 5, 2]])

Проверим случайную матрицу на обратимость.

In [19]:
np.linalg.inv(random_matrix)

array([[ 0.06374502,  0.33466135, -0.38645418,  0.04780876],
       [-0.24701195,  0.20318725, -0.12749004,  0.31474104],
       [ 0.37450199, -0.53386454,  0.35458167, -0.21912351],
       [-0.29083665,  0.22310757,  0.07569721,  0.03187251]])

Ошибки не возникло, значит матрица обратима

Умножим матрицу признаков features_matrix на матрицу random_matrix. Результат сохраним в переменной encoded_matrix.

In [20]:
encoded_matrix = features_matrix @ random_matrix
encoded_matrix[:5, :]

array([[248297, 198533, 397100, 397135],
       [190325, 152144, 304327, 304370],
       [105203,  84087, 168203, 168232],
       [208653, 166875, 333757, 333772],
       [130703, 104488, 209004, 209029]])

Данные потеряли свои оригинальные значения, теперь в них хранится непонятное число

In [21]:
type(encoded_matrix)

numpy.ndarray

*Данные преобразовали. Можно приступить к обучению модели*

#### Качество модели с исходными признаками

Разделим данные на обучающие и тестовые

In [22]:
features_train, features_test, target_train, target_test = train_test_split(
    encoded_matrix, target, test_size=0.25, random_state=12345)

In [23]:
print('Размер тренировочной выборки', features_train.shape)
print('Размер тестовой выборки', features_test.shape)

Размер тренировочной выборки (3750, 4)
Размер тестовой выборки (1250, 4)


In [24]:
model = LinearRegression()
model.fit(features_train, target_train)

R2_encoded_matrix = r2_score(target_test, model.predict(features_test))

print("Коэффициенты - вектор w:",model.coef_)
print("R2 =", R2_encoded_matrix)

Коэффициенты - вектор w: [ 0.01249466 -0.00114101 -0.00959304  0.00235431]
R2 = 0.43522756840807486


#### Качество модели с отмасштабированными признаками

In [25]:
regressor = LinearRegression()
scaller = StandardScaler()
scaller.fit(features_train, target_train)

pipeline = Pipeline([("standard_scaller", scaller),("linear_regression", regressor)])
pipeline.fit(features_train, target_train)

R2_encoded_matrix_scaled = r2_score(target_test, pipeline.predict(features_test))

print("R2 =", R2_encoded_matrix_scaled)

R2 = 0.4352275684077155


Сравним показатели качества моделей

In [26]:
result = pd.DataFrame(data= [R2_LR_origin_data_scaled,
                      R2_origin_data,
                      R2_encoded_matrix,
                      R2_encoded_matrix_scaled], 
                     columns=['R2'], 
                     index=['Линейная регрессия',
                            'Линейная регрессия c масштабом',
                            'Линейная регрессия на преобразованных признаках',
                            'Линейная регрессия на преобразованных признаках c масштабом',])
result

R2
Линейная регрессия                                  0.435228
Линейная регрессия c масштабом                      0.435228
Линейная регрессия на преобразованных признаках     0.435228
Линейная регрессия на преобразованных признаках...  0.435228

Метрики идентичны, а это значит, что качество модели линейной регрессии после преобразования данных не изменилось

## Итоговый вывод

В ходе работы было проделано:
- Загружены и изучены данные
- Математически доказано, что качество линейной регрессии не изменится, если умножать признаки на обратимую матрицу
- Создан алгоритм преобразования данных
- Протестирован алгоритм преобразования данных и проверена метрика R2 для данных без преобразования и с ним

По результатам работы использования матричных операций можно увидеть, что данные очень просто зашифровать от распознования, имея правильную матрицу.